<a href="https://colab.research.google.com/github/priyarane879/Employee-Attrition-Analysis/blob/main/SQL_Capstone_1_Part_2(Priya_Naik).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Capstone 1 - Part 2

In the first part of the capstone, we focused on Data Retrieval, Data Preprocessing, Feature Engineering and Exploratory Data Analysis using Python & its libraries. Now we are going to shift gears and gain insights into our HR Analytics dataset using SQL.

## TODO: Make use of SQL to do the following:

### Create a SQLITE3 DB using the CSV file (2 pts). Please refer this [link](https://www.linkedin.com/pulse/accessing-sqlite3-database-from-jupyter-notebook-using-varun-lobo/) and this [link](https://www.geeksforgeeks.org/how-to-import-csv-file-in-sqlite-database-using-python/) to know more.

### Calculate the Attrition Rate and summarize attrition (3 pts) by:
- Gender
- Department
- Age
- Average monthly income by job level
- Years at company

### Continue using SQL to explore main reasons for attrition (3 pts), For example:
- Why do more people over 50 years old leave the company than people who aged 40-50?
- Why do people with higher pay still leave the company?
- Which factors drive employees who work at company less than 5 years to leave?

### Effective Communication (2 pts)
- Please make use of markdown cells to communicate your thought process, why did you think of performing a step? what was the observation from the query? etc.
- The code should be commented so that it is readable for the reviewer.

### Grading and Important Instructions
- Each of the above steps are mandatory and should be completed in good faith
- Make sure before submitting that the code is in fully working condition
- It is fine to make use of ChatGPT, stackoverflow type resources, just provide the reference links from where you got it
- Debugging is an art, if you find yourself stuck with errors, take help of stackoverflow and ChatGPT to resolve the issue and if it's still unresolved, reach out to me for help.
- You need to score atleast 7/10 to pass the project, anything less than that will be marked required, needing resubmission.
- Feedback will be provided on 3 levels (Awesome, Suggestion, & Required). Required changes are mandatory to be made.
- For submission, please upload the project on github and share the link to the file with us through LMS.

#### Calculate Attrition Rate


In [1]:
import csv
import sqlite3
import pandas as pd
import sql


In [2]:
!pip install db-sqlite3
!pip install pandasql


  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 172.0/172.0 kB 3.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for db-sqlite3: filename=db_sqlite3-0.0.1-py3-none-any.whl size=1769 sha256=11c78772bdffbe0ef6a6bc6e499c7c1a56013b1cdb9b80a49713a2ec505e0f11
  Stored in directory: /root/.cache/pip/wheels/d2/76/f8/d078e1609571cb2d847296a2bb130a355498554aa647e957ba
  Created wheel for db: filename=db-0.1.1-py3-none-any.whl size=3874 sha256=79377faf33bc03b02946810be1a18a8f099c38b6ca691525fb06663846ff292d
  Stored in directory: /root/.cache/pip/wheels/1f/a9/df/8e343ae1489f35fa96888200ee3340217512b98bed2c2def2d
  Created wheel for antiorm: filename=antiorm-1.2.1-py3-none-any.whl size=31665 sha256=b840b17b72c669a889cfba55559dc6b083391b995c9a0828c88e1d0d72d390c6
  Stored in directory: /root/.cache/pip/wheels/71/04/d9/c9b4297ae902eb961bc1cff29b248a7e706693200ef2656fb2
Successfully built db-sq

In [3]:
### Implement code
#connecting with the google drive for reading the file from drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
'EnvironmentSatisfaction','OverTime', 'PercentSalaryHike', 'PerformanceRating',\
       'RelationshipSatisfaction','JobSatisfaction','MonthlyIncome','WorkLifeBalance','NumCompaniesWorked','PerformanceRating','YearsSinceLastPromotion'

('EnvironmentSatisfaction',
 'OverTime',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'JobSatisfaction',
 'MonthlyIncome',
 'WorkLifeBalance',
 'NumCompaniesWorked',
 'PerformanceRating',
 'YearsSinceLastPromotion')

In [5]:
try:

    # Import csv and extract data
    with open('/content/drive/MyDrive/Elevate Pro/SQL/HR-Analytics.csv', 'r') as fin:
        dr = csv.DictReader(fin)



        employee_info = [(i['Attrition'], i['BusinessTravel'],i['\ufeffAge'],i['DailyRate'],i['Department'],i['DistanceFromHome'],i['Education'],
                          i['EducationField'], i['EnvironmentSatisfaction'],i['Gender'],i['HourlyRate'],i['JobInvolvement'],i['JobLevel'],
                          i['JobRole'],i['JobSatisfaction'], i['MaritalStatus'],i['MonthlyIncome'],
                          i['MonthlyRate'],i['NumCompaniesWorked'],i['Over18'],i['OverTime'],i['PercentSalaryHike'],i['PerformanceRating'],
                          i['RelationshipSatisfaction'],i['StandardHours'],i['StockOptionLevel'],i['TotalWorkingYears'],i['TrainingTimesLastYear'],
                          i['WorkLifeBalance'],i['YearsAtCompany'],i['YearsInCurrentRole'],i['YearsSinceLastPromotion'],i['YearsWithCurrManager']
                          )for i in dr]
        print(employee_info)

    # Connect to SQLite
    sqliteConnection = sqlite3.connect('sql.db')
    #sqliteConnection = sqlite3.connect('sql.db')
    cursor = sqliteConnection.cursor()

    # Create student table
    # cursor.execute('create table table_HR(Gender varchar2(10),Department varchar2(10), \
    # Age int,MonthlyIncome int,YearsAtCompany int,Attrition Boolean(10),JobLevel int);')
    cursor.execute('create table table_HR( Age INT, Attrition VARCHAR(3), BusinessTravel VARCHAR(50), DailyRate int, \
    Department VARCHAR(100),DistanceFromHome INT, Education INT, EducationField VARCHAR(100), \
    EnvironmentSatisfaction INT, Gender VARCHAR(10),  HourlyRate INT, JobInvolvement INT, JobLevel INT, JobRole VARCHAR(100),\
    JobSatisfaction INT, MaritalStatus VARCHAR(20), MonthlyIncome INT, MonthlyRate INT, NumCompaniesWorked INT, Over18 VARCHAR(3),OverTime VARCHAR(3),\
    PercentSalaryHike INT,PerformanceRating INT, RelationshipSatisfaction INT,  StandardHours INT, StockOptionLevel INT, TotalWorkingYears INT,\
    TrainingTimesLastYear INT, WorkLifeBalance INT, YearsAtCompany INT, YearsInCurrentRole INT, YearsSinceLastPromotion INT, YearsWithCurrManager INT);')

    # Insert data into table
    cursor.executemany(
        "insert into table_HR ('Attrition', 'BusinessTravel', 'Age','DailyRate', 'Department',\
       'DistanceFromHome', 'Education', 'EducationField',\
        'EnvironmentSatisfaction', 'Gender', 'HourlyRate',\
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',\
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',\
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',\
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',\
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',\
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',\
       'YearsWithCurrManager') VALUES (?, ?,?,?,?,?,?,?, ?,?,?,?,?,?,?, ?,?,?,?,?,?,?, ?,?,?,?,?,?,?, ?,?,?,?);", employee_info)

    # Show student table
    print("done with data")
    cursor.execute('select * from table_HR;')

    # View result
    result = cursor.fetchall()
    print(result)

    # Commit work and close connection
    sqliteConnection.commit()
    cursor.close()

except sqlite3.Error as error:
    print('Error occurred - ', error)

finally:
    if sqliteConnection:
        sqliteConnection.close()
        print('SQLite Connection closed')

[('Yes', 'Travel_Rarely', '41', '1102', 'Sales', '1', '2', 'Life Sciences', '2', 'Female', '94', '3', '2', 'Sales Executive', '4', 'Single', '5993', '19479', '8', 'Y', 'Yes', '11', '3', '1', '80', '0', '8', '0', '1', '6', '4', '0', '5'), ('No', 'Travel_Frequently', '49', '279', 'Research & Development', '8', '1', 'Life Sciences', '3', 'Male', '61', '2', '2', 'Research Scientist', '2', 'Married', '5130', '24907', '1', 'Y', 'No', '23', '4', '4', '80', '1', '10', '3', '3', '10', '7', '1', '7'), ('Yes', 'Travel_Rarely', '37', '1373', 'Research & Development', '2', '2', 'Other', '4', 'Male', '92', '2', '1', 'Laboratory Technician', '3', 'Single', '2090', '2396', '6', 'Y', 'Yes', '15', '3', '2', '80', '0', '7', '3', '3', '0', '0', '0', '0'), ('No', 'Travel_Frequently', '33', '1392', 'Research & Development', '3', '4', 'Life Sciences', '4', 'Female', '56', '3', '1', 'Research Scientist', '3', 'Married', '2909', '23159', '1', 'Y', 'Yes', '11', '3', '3', '80', '0', '8', '3', '3', '8', '7', '3',

In [6]:
sqliteConnection = sqlite3.connect('sql.db')
    #sqliteConnection = sqlite3.connect('sql.db')
cursor = sqliteConnection.cursor()


In [7]:
pd.read_sql_query('''SELECT * FROM table_HR''',sqliteConnection)

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                    1          2  Life Sciences                        2   
1                    8          1  Life Sciences                        3   
2                    2          2          Other                        4   
3                    3          4  Life Sciences                        4   
4                    2          1        Medical                        1   
...                ...        ...            ...                      ...   
1465                23          2        Medical                        3   
1466                 6          1        Medical                        4   
1467                 4          3  Life Sciences                        2   
1468                 2          3        Medical                        4   
1469                 8          3        Medical                        2   

      Gender  ...  RelationshipSatisfaction  StandardHours  StockOptionLevel  \
0     Female  ...                         1             80                 0   
1       Male  ...                         4             80                 1   
2       Male  ...                         2             80                 0   
3     Female  ...                         3             80                 0   
4       Male  ...                         4             80                 1   
...      ...  ...                       ...            ...               ...   
1465    Male  ...                         3             80                 1   
1466    Male  ...                         1             80                 1   
1467    Male  ...                         2             80                 1   
1468    Male  ...                         4             80                 0   
1469    Male  ...                         1             80                 0   

     TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                    8                      0               1               6   
1                   10                      3               3              10   
2                    7                      3               3               0   
3                    8                      3               3               8   
4                    6                      3               3               2   
...                ...                    ...             ...             ...   
1465                17                      3               3               5   
1466                 9                      5               3               7   
1467                 6                      0               3               6   
1468                17                      3               2               9   
1469                 6                      3               4               4   

      YearsInCurrentRole  YearsSinceLastPromotion YearsWithCurrManager  
0                      4                        0                    5  
1                      7                        1                    7  
2  

In [8]:
pd.read_sql_query('''SELECT Attrition,ROUND((COUNT(*) * 100.0 / SUM(COUNT(*)) OVER ()), 1)
                    AS Attrition_rate FROM table_HR GROUP BY Attrition order by Attrition_rate ''',sqliteConnection)

Attrition  Attrition_rate
0       Yes            16.1
1        No            83.9

#### Sample Output

![image](https://api-v4.skyprepapp.com/public_api/da/857886?view_key=eyJvYmplY3RfdHlwZSI6ImFydGljbGUiLCJvYmplY3RfaWQiOjg1Nzg4NiwiaG1hYyI6ImUyNGU0YWRhNWQwMDkxNWE2NWNiY2EwNGFlNDNiMTBlYmYyNWY0YzAiLCJleHBpcmVzX2F0Ijo0ODkxMzgxMTk5LCJvdHRfdG9rZW4iOm51bGx9)

Observation:-

The table presents attrition data with two categories:

1. 16.1% of employees have left the organization (attrition = true).
2. 83.9% of employees have remained with the organization (attrition = false).

This indicates that the majority of employees are retained, while a smaller portion has exited.

#### Find Attrition by Gender

In [9]:
### Implement code
pd.read_sql_query('''SELECT Attrition, Gender, COUNT(*) as Count_gender,  ROUND((COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (partition by Gender)), 1)
                    AS Attrition_by_gender FROM table_HR GROUP BY Attrition, Gender order by Attrition_by_gender ''',sqliteConnection)

Attrition  Gender  Count_gender  Attrition_by_gender
0       Yes  Female            87                 14.8
1       Yes    Male           150                 17.0
2        No    Male           732                 83.0
3        No  Female           501                 85.2

Observations:

1. Males have a slightly higher attrition rate (17.0%) compared to females (14.8%).
2. The majority of both genders remain with the organization.
More males have left in absolute numbers (150) compared to females (87), likely due to a larger male employee population.








#### Sample output

![image](https://api-v4.skyprepapp.com/public_api/da/857882?view_key=eyJvYmplY3RfdHlwZSI6ImFydGljbGUiLCJvYmplY3RfaWQiOjg1Nzg4MiwiaG1hYyI6IjM1NDU5N2E4MjZmMTMxNWFhMTA4NzFiYTFkZjQ5YjBhNTYyOWYwMzgiLCJleHBpcmVzX2F0Ijo0ODkxMzgxMTk5LCJvdHRfdG9rZW4iOm51bGx9)

#### Find Attrition by Dept

In [10]:
#### Find Attrition by Gender### Implement code
pd.read_sql_query('''SELECT Department, Attrition, Count(*) as Department_attrition
                    FROM table_HR GROUP BY Department, Attrition order by Department''',sqliteConnection)

Department Attrition  Department_attrition
0         Human Resources        No                    51
1         Human Resources       Yes                    12
2  Research & Development        No                   828
3  Research & Development       Yes                   133
4                   Sales        No                   354
5                   Sales       Yes                    92

Observations:

1. The Research & Development department has the highest number of employees (both retained and attrited), with 133 employees leaving.
2. The Sales department has 92 employees leaving, indicating a relatively high attrition count.
3. The Human Resources department has the lowest attrition in absolute numbers (12 employees).
Across all departments, the majority of employees remain with the company.

#### Sample Output

<img src="https://api-v4.skyprepapp.com/public_api/da/857884?view_key=eyJvYmplY3RfdHlwZSI6ImFydGljbGUiLCJvYmplY3RfaWQiOjg1Nzg4NCwiaG1hYyI6IjdmOTA4NzJjODZmYTgxOGUyMzkyYTBlZjhjYjljNjA4ZGM3NjkzMWYiLCJleHBpcmVzX2F0Ijo0ODkxMzgxMTk5LCJvdHRfdG9rZW4iOm51bGx9" width="450" height="450">

#### Find Attrition by Age Groups

In [11]:
pd.read_sql_query('''SELECT Attrition, CASE
         WHEN Age < 30 THEN 'Under 30'
         WHEN Age BETWEEN 30 AND 39 THEN '30-40'
         WHEN Age BETWEEN 40 AND 49 THEN '40-50'
         ELSE 'Over 50'
         END AS Age_group,
         COUNT(*) as num,
         ROUND((COUNT(*) * 100.0 / SUM(COUNT(*))
         OVER (PARTITION  by CASE
                              WHEN Age < 30 THEN 'Under 30'
                              WHEN Age BETWEEN 30 AND 39 THEN '30-40'
                              WHEN Age BETWEEN 40 AND 49 THEN '40-50'
                              ELSE 'Over 50'
                              END )),2) as percent_by_age
         FROM table_HR GROUP BY Attrition, CASE
                                         WHEN Age < 30 THEN 'Under 30'
                                         WHEN Age BETWEEN 30 AND 39 THEN '30-40'
                                         WHEN Age BETWEEN 40 AND 49 THEN '40-50'
                                         ELSE 'Over 50'
                                         END order by percent_by_age desc''',sqliteConnection)

Attrition Age_group  num  percent_by_age
0        No     40-50  315           90.26
1        No   Over 50  150           86.71
2        No     30-40  533           85.69
3        No  Under 30  235           72.09
4       Yes  Under 30   91           27.91
5       Yes     30-40   89           14.31
6       Yes   Over 50   23           13.29
7       Yes     40-50   34            9.74

Observation:-
1. Younger employees (Under 30) are the most likely to leave, possibly due to career exploration, job dissatisfaction, or better opportunities.

2. Middle-aged employees (30-50) are more stable, likely due to career growth, financial commitments, or job satisfaction.

3. Older employees (Over 50) have low attrition, possibly due to job security, nearing retirement, or fewer external job opportunities.

#### Sample Output

![image](https://api-v4.skyprepapp.com/public_api/da/857885?view_key=eyJvYmplY3RfdHlwZSI6ImFydGljbGUiLCJvYmplY3RfaWQiOjg1Nzg4NSwiaG1hYyI6IjcwNGVlOWIwYzg1MmMwNmNhODg0NmYyZDFlNWE3OTU1MTFhMGVmYWYiLCJleHBpcmVzX2F0Ijo0ODkxMzgxMTk5LCJvdHRfdG9rZW4iOm51bGx9)

#### Find Attrition by Monthly Income

In [12]:
### Implement code
pd.read_sql_query('''SELECT Department, JobLevel, Round(Avg(MonthlyIncome),1) as avg_income,
                     ROUND(AVG(CASE WHEN Attrition = 'Yes' THEN MonthlyIncome END), 1) AS attrition_avg_income,
                     ROUND(AVG(CASE WHEN Attrition = 'Yes' THEN MonthlyIncome END) - AVG(MonthlyIncome), 1) AS difference
                    FROM table_HR
                    GROUP BY Department, JobLevel
                    HAVING AVG(CASE WHEN Attrition = 'Yes' THEN MonthlyIncome END) IS NOT NULL
                    order by Department''',sqliteConnection)

Department  JobLevel  avg_income  attrition_avg_income  \
0          Human Resources         1      2733.2                2415.7   
1          Human Resources         3      9623.0               10216.0   
2   Research & Development         1      2840.1                2687.4   
3   Research & Development         2      5291.2                5372.0   
4   Research & Development         3     10170.5                9503.8   
5   Research & Development         4     15634.7               12169.0   
6   Research & Development         5     19218.5               19550.0   
7                    Sales         1      2506.7                2373.4   
8                    Sales         2      5746.1                5917.0   
9                    Sales         3      9282.3                9202.8   
10                   Sales         4     15166.3               13395.8   
11                   Sales         5     19088.0               19334.5   

    difference  
0       -317.5  
1        593.0  
2       -152.7  
3         80.8  
4       -666.6  
5      -3465.7  
6        331.5  
7       -133.3  
8        170.9  
9        -79.5  
10     -1770.5  
11       246.5

1. Lower-level employees tend to leave with lower salaries than their department's average. This could indicate dissatisfaction with pay at the entry-level.

2. Mid-level employees in some departments experience salary drops upon leaving, possibly due to job market conditions or role adjustments.

3. Higher-level employees (Job Level 5) leaving tend to have competitive salaries, suggesting that attrition at senior levels might be due to better opportunities rather than dissatisfaction.

4. The largest negative income differences are in R&D (Job Level 4) and Sales (Job Level 4), indicating that higher-level attrition in these departments might come with pay cuts.

#### Sample Output

![image](https://api-v4.skyprepapp.com/public_api/da/857883?view_key=eyJvYmplY3RfdHlwZSI6ImFydGljbGUiLCJvYmplY3RfaWQiOjg1Nzg4MywiaG1hYyI6Ijg3NTU4ZDU1ZjRjN2U1YWI3ODQzYjM1NzFkNjBjMjEwNGY5NWI5ODUiLCJleHBpcmVzX2F0Ijo0ODkxMzgxMTk5LCJvdHRfdG9rZW4iOm51bGx9)

#### Find Attrition by Years At Company

In [13]:
### Implement code
pd.read_sql_query('''SELECT
    CASE
        WHEN YearsAtCompany < 2 THEN 'New Hires'
        WHEN YearsAtCompany BETWEEN 2 AND 5 THEN '2-5 years'
        WHEN YearsAtCompany BETWEEN 6 AND 10 THEN '6-10 years'
        WHEN YearsAtCompany BETWEEN 11 AND 20 THEN '11-20 years'
        ELSE 'Over 20 years'
    END AS tenure_years,
    COUNT(*) AS num,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER(), 1) AS percent
FROM table_HR where Attrition = 'Yes'
GROUP BY
    CASE
        WHEN YearsAtCompany < 2 THEN 'New Hires'
        WHEN YearsAtCompany BETWEEN 2 AND 5 THEN '2-5 years'
        WHEN YearsAtCompany BETWEEN 6 AND 10 THEN '6-10 years'
        WHEN YearsAtCompany BETWEEN 11 AND 20 THEN '11-20 years'
        ELSE 'Over 20 years'
    END
ORDER BY num DESC''',sqliteConnection)

tenure_years  num  percent
0      2-5 years   87     36.7
1      New Hires   75     31.6
2     6-10 years   55     23.2
3    11-20 years   12      5.1
4  Over 20 years    8      3.4

Observation:-
1. Early-career employees (New Hires & 2-5 years) are the most likely to leave, indicating that retention strategies should focus on engagement, career growth, and satisfaction within the first 5 years.

2. Mid-career employees (6-10 years) still experience some attrition, but the rate declines significantly, suggesting they find more stability.

3. Long-tenured employees (11+ years) have very low attrition, meaning that those who stay beyond a decade are likely to remain long-term.

#### Sample Output

![image](https://api-v4.skyprepapp.com/public_api/da/857881?view_key=eyJvYmplY3RfdHlwZSI6ImFydGljbGUiLCJvYmplY3RfaWQiOjg1Nzg4MSwiaG1hYyI6IjFhYmY3NGI4MzQ1NzViMWZkNDJlMjcwYTUyOTQ0OWQwZjJjMjhmNWUiLCJleHBpcmVzX2F0Ijo0ODkxMzgxMTk5LCJvdHRfdG9rZW4iOm51bGx9)

#### Continue using SQL to explore main reasons for attrition (3 pts), For example:
- Why do more people over 50 years old leave the company than people who aged 40-50?
- Why do people with higher pay still leave the company?
- Which factors drive employees who work at company less than 5 years to leave?

In [14]:
pd.read_sql_query('''SELECT  CASE
                            WHEN Age BETWEEN 40 AND 50 THEN '40-50'
                            ELSE 'Over 50'
                        END AS Age_group,OverTime,
                        AVG(EnvironmentSatisfaction),
                        AVG(JobSatisfaction),
                        AVG(PercentSalaryHike),
                        AVG(PerformanceRating),
                        AVG(RelationshipSatisfaction),
                        AVG(MonthlyIncome),
                        AVG(WorkLifeBalance),
                        AVG(NumCompaniesWorked),
                        AVG(YearsSinceLastPromotion)
                    FROM table_HR where Attrition='Yes'
                    GROUP BY  Age_group, OverTime
                    ''', sqliteConnection)

Age_group OverTime  AVG(EnvironmentSatisfaction)  AVG(JobSatisfaction)  \
0     40-50       No                      2.000000              2.352941   
1     40-50      Yes                      2.500000              2.545455   
2   Over 50       No                      2.387097              2.365591   
3   Over 50      Yes                      2.600000              2.561905   

   AVG(PercentSalaryHike)  AVG(PerformanceRating)  \
0               15.058824                3.176471   
1               13.818182                3.090909   
2               14.752688                3.118280   
3               15.676190                3.200000   

   AVG(RelationshipSatisfaction)  AVG(MonthlyIncome)  AVG(WorkLifeBalance)  \
0                       2.823529         7708.000000              2.823529   
1                       2.909091         6377.500000              2.454545   
2                       2.462366         4385.903226              2.591398   
3                       2.619048         4336.295238              2.733333   

   AVG(NumCompaniesWorked)  AVG(YearsSinceLastPromotion)  
0                 3.941176                      3.588235  
1                 3.318182                      1.863636  
2                 2.795699                      1.709677  
3                 2.828571                      1.904762

Observation:-
1. Employees over 50 (No Overtime) earn 4,385.90, while those aged 40-50 (No Overtime) earn 7,708.00. Even though older employees received higher salary hikes (15.67%), their base salaries remain significantly lower.
2. Employees over 50 working overtime have an even lower work-life balance, which could lead to burnout. Employees aged 40-50 (No Overtime) have a higher job satisfaction score (2.35) compared to over 50 (No Overtime) at 2.36.
 Over 50 (Overtime) have slightly better job satisfaction (2.56) but still not significantly high.
3. Employees aged 40-50 (No Overtime) have not been promoted in ~3.59 years, whereas those over 50 (No Overtime) have not been promoted in ~1.7 years.
This suggests that even though promotions happen, they do not result in significant salary increases, which may lead to frustration.

4. Employees aged 40-50 have better relationship satisfaction (2.82-2.90) than those over 50 (2.46-2.61).Performance ratings are similar (3.17 for 40-50 vs. 3.15 for over 50), meaning older employees are performing well but may feel undervalued.

Q.2 Why do people with higher pay still leave the company?

In [30]:
pd.read_sql_query('''
       WITH IncomePercentiles AS (
    SELECT
        MonthlyIncome,
        NTILE(4) OVER (ORDER BY MonthlyIncome) AS IncomeGroup
    FROM table_HR
    WHERE MonthlyIncome > (SELECT AVG(MonthlyIncome) FROM table_HR)
)
SELECT
    CASE
        WHEN ip.IncomeGroup = 1 THEN 'Low (0-25th Percentile)'
        WHEN ip.IncomeGroup = 2 THEN 'Medium (25th-50th Percentile)'
        WHEN ip.IncomeGroup = 3 THEN 'High (50th-75th Percentile)'
        WHEN ip.IncomeGroup = 4 THEN 'Very High (75th-100th Percentile)'
    END AS IncomeCategory,
    t.Attrition,
    Avg(t.MonthlyIncome),
    Avg(t.JobSatisfaction),
    Avg(t.WorkLifeBalance),
    Avg(t.YearsSinceLastPromotion),
    Avg(t.PerformanceRating)

FROM table_HR t
JOIN IncomePercentiles ip ON t.MonthlyIncome = ip.MonthlyIncome
WHERE t.MonthlyIncome > (SELECT AVG(MonthlyIncome) FROM table_HR)
GROUP BY IncomeCategory, t.Attrition;
                    ''', sqliteConnection)

IncomeCategory Attrition  Avg(t.MonthlyIncome)  \
0        High (50th-75th Percentile)        No          12698.710526   
1        High (50th-75th Percentile)       Yes          12175.727273   
2            Low (0-25th Percentile)        No           7201.870690   
3            Low (0-25th Percentile)       Yes           7352.357143   
4      Medium (25th-50th Percentile)        No           9521.038462   
5      Medium (25th-50th Percentile)       Yes           9762.840000   
6  Very High (75th-100th Percentile)        No          18177.721311   
7  Very High (75th-100th Percentile)       Yes          19463.800000   

   Avg(t.JobSatisfaction)  Avg(t.WorkLifeBalance)  \
0                2.736842                2.824561   
1                2.636364                2.181818   
2                2.887931                2.775862   
3                2.500000                2.785714   
4                2.769231                2.778846   
5                2.360000                2.800000   
6                2.696721                2.762295   
7                2.400000                3.000000   

   Avg(t.YearsSinceLastPromotion)  Avg(t.PerformanceRating)  
0                        3.166667                  3.184211  
1                        7.363636                  3.000000  
2                        1.836207                  3.094828  
3                        1.642857                  3.142857  
4                        3.759615                  3.153846  
5                        3.560000                  3.120000  
6                        4.786885                  3.147541  
7                        7.200000                  3.000000

Observation:-
1. Employees who did not leave have slightly higher average salaries than those who left.
2. Employees with lower job satisfaction are more likely to leave, regardless of salary.
3. Poor Work-Life Balance contributes to attrition, especially in High & Low percentiles.
4. Long promotion gaps strongly correlate with attrition.Even high salaries do not retain employees if promotions are delayed.
5.  Employees with similar performance ratings still leave, indicating attrition is driven more by promotion delays & job satisfaction than performance.

Observation

Q.3) Which factors drive employees who work at company less than 5 years to leave?

In [ ]:
pd.read_sql_query('''
SELECT
    YearsAtCompany,Attrition,
    COUNT(*) AS Number_of_Employees,
    AVG(MonthlyIncome) AS Average_Monthly_Income,
    AVG(JobLevel) AS Average_JobLevel,
    AVG(JobSatisfaction) AS Average_JobSatisfaction,
    AVG(WorkLifeBalance) AS Average_WorkLifeBalance
FROM table_HR
WHERE YearsAtCompany < 5
GROUP BY YearsAtCompany,Attrition
ORDER BY Attrition
''', sqliteConnection)


YearsAtCompany Attrition  Number_of_Employees  Average_Monthly_Income  \
0               0        No                   28             4715.357143   
1               1        No                  112             5479.991071   
2               2        No                  100             5168.460000   
3               3        No                  108             5326.388889   
4               4        No                   91             5361.780220   
5               0       Yes                   16             3060.250000   
6               1       Yes                   59             3165.152542   
7               2       Yes                   27             4157.407407   
8               3       Yes                   20             3743.100000   
9               4       Yes                   19             3903.368421   

   Average_JobLevel  Average_JobSatisfaction  Average_WorkLifeBalance  
0          1.678571                 2.785714                 2.750000  
1          1.776786                 2.821429                 2.812500  
2          1.680000                 2.950000                 2.790000  
3          1.740741                 2.703704                 2.777778  
4          1.802198                 2.582418                 2.670330  
5          1.250000                 2.250000                 2.750000  
6          1.305085                 2.525424                 2.694915  
7          1.333333                 2.592593                 2.518519  
8          1.350000                 2.000000                 2.950000  
9          1.421053                 2.684211                 2.421053

Observation:-
1. Employees who left had significantly lower average monthly income and job levels than those who stayed.
2. Employees who left had lower work-life balance scores.
3. New hires with lower salaries are more likely to leave early.
